Keep archived out of VimPerformanceState

kbrock opened the pull request:
reference:
- https://github.com/ManageIQ/manageiq/issues/22593

requires:
- [x] https://github.com/ManageIQ/manageiq/pull/22594
- [x] https://github.com/ManageIQ/manageiq/pull/22589 (first 4 commits)

extracted:
- [x] https://github.com/ManageIQ/manageiq/pull/22611 (not required but related)

Overview
========

We were putting too many archived container ids in the `VimPerformanceState` records.
This was filling up the database, causing us to process too many records, slowing down metrics capture, and slowing down metrics rollups.

Before
======

- Don't take into account whether a record was archived.

After
=====

- We are no longer putting the ids for records that were archived before our time window in the performance state records

Numbers
=======

|       ms |       bytes |   objects |query |  qry ms |     rows |` comments`
|      ---:|         ---:|       ---:|  ---:|     ---:|      ---:|  ---
|  2,131.6 | 77,558,748* | 1,070,108 |   13 | 1,760.5 |   13,680 |`rollup-create-vps-before`
|  1,632.5 |  5,139,458* |   393,268 |   13 | 1,517.7 |      174 |`rollup-create-vps-after`
|     23%  |         93% |       63% |      |   13.8% |      99% | diff

```
* Memory usage does not reflect 175,949 freed objects.
* Memory usage does not reflect 59,056 freed objects.
```

- this example was chosen because of the high number of archived records
- this example created an VimPerformanceState record. otherwise, there would have been no change
- Estimated bytes total changed to 90311091 and 5911237 respectively. This resulted in the same bytes savings
- The query ms savings is very dependent upon the cache status, and was not stable during the tests. But rows returned does give an indicator that the work on the database was reduced significantly
- the number of rows returned is the difference between bringing back the archived records for creating the VPS and not.
- The size of the VPS#state_data went from 167,234 => 1,752 characters. The record is more than 150x smaller.

Outstanding
===========

- migration/script to prune out existing VimPerformanceStates that contain archived ids

Fryguy commented on app/models/vim_performance_state.rb:
```diff
@@ -4,6 +4,11 @@ class VimPerformanceState < ApplicationRecord
   serialize :state_data
 
   belongs_to :resource, :polymorphic => true
+  default_value_for :capture_interval, 3600
```
I'm so bad at recognizing numbers - prefer at least having a comment, or perhaps use `1.hour.to_i`

```suggestion
  default_value_for :capture_interval, 3600 # 1 hour
```

Fryguy commented on spec/models/mixins/archived_mixin_spec.rb:
```diff
@@ -0,0 +1,83 @@
+RSpec.describe ArchivedMixin do
+  let(:model_name) { :container }
+  let(:klass) { Container }
+  # let(:model) { klass.new }
+  # let(:archived_model) { klass.new() }
```
Commented?

Fryguy commented on app/models/vim_performance_state.rb:
```diff
@@ -222,6 +224,25 @@ def capture_assoc_ids
+    records
+  end
+
+
```
Super-mega-extremenly-minor bonus extra line

Fryguy commented on spec/models/mixins/archived_mixin_spec.rb:
```diff
@@ -0,0 +1,83 @@
+  end
+
+  describe ".active_for" do
+    it "detects active are valid" do
+      active_model
+      expect(klass.active_for((2.years.ago)..(1.year.ago))).to eq([active_model])
+    end
+
+    it "detects not yet deleted records are valid" do
+      archived_model
+      expect(klass.active_for((2.years.ago)..(1.year.ago))).to eq([archived_model])
+    end
+
+    it "detects already deleted records are not valid" do
```
valid/invalid terminology sounds strange here - the records aren't "bad".  I think prefer in range / out of range

```suggestion
    it "detects active records in range" do
      active_model
      expect(klass.active_for((2.years.ago)..(1.year.ago))).to eq([active_model])
    end

    it "detects not yet deleted records are in range" do
      archived_model
      expect(klass.active_for((2.years.ago)..(1.year.ago))).to eq([archived_model])
    end

    it "detects already deleted records are out of range" do
```

Fryguy commented on spec/models/mixins/archived_mixin_spec.rb:
```diff
@@ -0,0 +1,83 @@
+    it "makes archived" do
+      expect(model.archived?).to be false
+      model.archive!
+      expect(model.archived?).to be true
```
Can you also add some checks about the value of deleted_on?

Fryguy replied:
Same with unarchive below.

Fryguy commented on spec/models/vim_performance_state_spec.rb:
```diff
@@ -1,5 +1,40 @@
+    it "uses now for Infra" do
+      hardware = FactoryBot.create(:hardware, :cpu_sockets => 2)
+      host = FactoryBot.create(:host, :hardware => hardware)
+      state = VimPerformanceState.capture(host, capture_time)
+
+      expect(state.timestamp).to be_within(1.minute).of(Time.now.beginning_of_hour)
```
This is a very weird condition (I fully understand that this is the way it works currently) and a future reader will not understand this.  I think it would be helpful to have a `NOTE` comment or something similar describing _why_ now is used for Infra and thus capture_time is ignored.  Not sure if that note should live here or in the code or both.

kbrock replied:
luckily we no longer need to deal with the nuance of infra vs containers

Fryguy commented on spec/models/vim_performance_state_spec.rb:
```diff
@@ -1,5 +1,40 @@
+  end
+
+  describe "#capture_assoc_ids" do
+    it "captures runing containers" do
```
typo 
```suggestion
    it "captures running containers" do
```

Fryguy commented on spec/models/vim_performance_state_spec.rb:
```diff
@@ -1,5 +1,40 @@
+    it "captures runing containers" do
+      container_image = FactoryBot.create(:container_image)
+      # not deleted is captured
+      containers = FactoryBot.create_list(:container, 1, :container_image => container_image)
```
Why is this using create_list? (Minor, but I just wasn't sure if you meant to only create one or were planning on more than one)

Fryguy replied:
(Similar question for the other create_list calls below)

kbrock replied:
yea, only creating one, (could create more) but using that over then turning this into an array

Fryguy replied:
Oh i see - it's to create an Array of 1.

jrafanie commented on app/models/vim_performance_state.rb:
```diff
@@ -46,22 +51,29 @@ class VimPerformanceState < ApplicationRecord
+
+    # Soft deletes can capture in the past (keep the timestamp), otherwise capture current state (timestamp = now)
+    ts = Time.now if ts.nil? || !objs.first.class.respond_to?(:active_for)
+    ts = ts.utc
```
FYI, existing `rollup_spec.rb` fails with 

```
     NoMethodError:
       undefined method `utc' for "2010-04-14T21:00:00Z":String
     # ./app/models/vim_performance_state.rb:64:in `capture'
     # ./app/models/metric/ci_mixin/state_finders.rb:34:in `vim_performance_state_for_ts'
     # ./app/models/metric/ci_mixin/state_finders.rb:82:in `containers_from_vim_performance_state_for_ts'
     # ./app/models/metric/rollup.rb:271:in `rollup_child_metrics'
     # ./app/models/metric/rollup.rb:211:in `block in rollup_hourly'
     # ./app/models/metric/rollup.rb:211:in `each'
     # ./app/models/metric/rollup.rb:211:in `rollup_hourly'
     # ./app/models/metric/ci_mixin/rollup.rb:96:in `block (2 levels) in perf_rollup'
     # /Users/joerafaniello/.gem/ruby/3.0.6/gems/more_core_extensions-4.4.0/lib/more_core_extensions/core_ext/benchmark/realtime_store.rb:20:in `realtime_store'
     # /Users/joerafaniello/.gem/ruby/3.0.6/gems/more_core_extensions-4.4.0/lib/more_core_extensions/core_ext/benchmark/realtime_store.rb:56:in `realtime_block'
     # ./app/models/metric/ci_mixin/rollup.rb:95:in `block in perf_rollup'
     # /Users/joerafaniello/.gem/ruby/3.0.6/gems/more_core_extensions-4.4.0/lib/more_core_extensions/core_ext/benchmark/realtime_store.rb:20:in `realtime_store'
     # /Users/joerafaniello/.gem/ruby/3.0.6/gems/more_core_extensions-4.4.0/lib/more_core_extensions/core_ext/benchmark/realtime_store.rb:62:in `realtime_block'
     # ./app/models/metric/ci_mixin/rollup.rb:82:in `perf_rollup'
     # ./spec/models/metric/ci_mixin/rollup_spec.rb:439:in `block (4 levels) in <main>'
     # /Users/joerafaniello/.gem/ruby/3.0.6/gems/webmock-3.18.1/lib/webmock/rspec.rb:37:in `block (2 levels) in <main>'
```

jrafanie replied:
It's doing `container_image.perf_rollup("2010-04-14T21:00:00Z", 'hourly')`

kbrock replied:
I went ahead and fixed a caller and added code to protect this.
This code has been moved into 22589 (and this is built upon that)

Fryguy commented on spec/models/vim_performance_state_spec.rb:
```diff
@@ -1,5 +1,40 @@
+      FactoryBot.create(:container, :container_image => container_image, :deleted_on => (capture_time - 2.hours))
+
+      state = VimPerformanceState.capture(container_image, capture_time)
+      expect(state.get_assoc(:containers).map(&:to_i).sort).to eq(containers.map(&:id).sort)        
```
```suggestion
      expect(state.get_assoc(:containers)).to match_array(containers)
```

kbrock replied:
this is comparing `to_i` to `id`, so I did need to keep that in there

Fryguy commented on spec/models/vim_performance_state_spec.rb:
```diff
@@ -20,6 +55,45 @@
+      # not deleted is captured
+      containers = FactoryBot.create_list(:container, 1, :container_image => container_image)
+
+      state_data = {:assoc_ids => {:containers => {:on => (containers.map(&:id) + [22, 23, 24, 25, 26, 27])}}}
```
What are the random numbers for?  If it's to get "other" ids, prefer building the id list from existing records, because a) this doesn't handle multi-region properly where the ids aren't in the 0 range and b) there's a chance the existing records could fall into the hardcoded range and fail the test.

kbrock replied:
put in a comment.

these represent ids that are no longer valid.
If container.id is within that range, then one of those ids will work, but the others will be thrown away. so there will be no false positives in here.

I could do `containers.last.id` and build from there? - but that seemed like a bunch of code that will create invalid ids just as easily as this does.

Fryguy replied:
`container.last.id` sounds reasonable - we do this in other specs as well.  We've hit problems in other places where we fake ids for records that either don't exist or shouldn't exist and hardcoded ids have always eventually caused problems.

Fryguy commented on spec/models/mixins/archived_mixin_spec.rb:
```diff
@@ -0,0 +1,83 @@
+    end
+  end
+
+  describe ".active_for" do
```
I think we need a few more cases (looking at my drawing that I shared the other day).

(created/deleted wording is relative to the time range in question (i.e. during means during the time range))

case | created | deleted | in range
--- | --- | --- | ---
a | before | before | ❌ 
b | before | during | ✅ 
c | before | after | ✅ 
d | before | not | ✅ 
e | during | during | ✅ 
f | during | after | ✅ 
g | during | not | ✅ 
h | after | after | ❌ 
i | after | not | ❌

kbrock replied:
The problem: it is very tricky to define `created` across all container models.

@Fryguy Do you want to define different `active_on` scopes for each record?

Fryguy replied:
We don't need created for the first pass as discussed - I think h and i were hard for that reason.

Fryguy replied:
> @Fryguy Do you want to define different active_on scopes for each record?

Maybe?  Do have a list of the record types that are "different"?  Or possibly we can normalize a different way - perhaps @agrare has some thoughts here.

kbrock commented:
update:

- this is now build upon #22589 
- dropped future and other sophisticated logic. only capturing for now.

kbrock commented:
update:
- fixed failing test ( `s/&klass/klass/`)
- updated performance numbers in git comment.

update:
- rubocop in spec: `Time.now.utc.beginning_of_hour`

kbrock commented:
update:
- updated specs to better show where the change happened around capture and now fails when you comment out the changed code.

kbrock commented:
update:
- rebased 

UN-WIP: this has reduced quite a bit in complexity

kbrock commented on app/models/vim_performance_state.rb:
```diff
@@ -231,6 +221,25 @@ def capture_assoc_ids
+    return false unless resource.respond_to?(method)
+
+    records = resource.send(method)
+    records = records.not_archived_before(timestamp) if records.try(:klass).respond_to?(:not_archived_before)
```
This is the only actual changed line

kbrock replied:
> Why do we have the `try` here?
> -- @Fryguy 

If the association is a virtual attribute, like it is for `MiqRegion`, then it will return an array that does not have a `klass`.

kbrock commented on app/models/vim_performance_state.rb:
```diff
@@ -231,6 +221,25 @@ def capture_assoc_ids
+             else
+               assoc
+             end
+    return false unless resource.respond_to?(method)
```
this was a minor change to return something so the parent knows we need to `next`

Fryguy commented:
Backported to `petrosian` in commit 10f15a3f1414f1210269dd00dcccd464e6cb64d8.

```text
commit 10f15a3f1414f1210269dd00dcccd464e6cb64d8
Author: Jason Frey <fryguy9@gmail.com>
Date:   Fri Aug 11 15:07:24 2023 -0400

    Merge pull request #22585 from kbrock/capture_vim_performance
    
    Keep archived out of VimPerformanceState
    
    (cherry picked from commit 900fb1bc6d235dc41668dc11cc79f0621429cc4f)
```
